File: app/src/main/java/com/neko/todolistapplication/MainActivity.java
package com.neko.todolistapplication;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.EditText;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private ArrayList<String> listOfTasks;
    private TaskItemAdapter taskAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.activity_main);



        listOfTasks = new ArrayList<String>();

        //long click on inside recycler

        TaskItemAdapter.OnLongClickListener olcl = new TaskItemAdapter.OnLongClickListener() {
            @Override
            public void onItemLongClicked(int position) {
                // remove item
                listOfTasks.remove(position);
                // notify adapter
                taskAdapter.notifyDataSetChanged();

                saveItems();
            }

        };

        loadItems();


        //recylcler stuff

        RecyclerView rcyView = findViewById(R.id.recyclerfield);
        taskAdapter = new TaskItemAdapter(listOfTasks,olcl);


        rcyView.setAdapter(taskAdapter);
        rcyView.setLayoutManager(new LinearLayoutManager(this));



        //button stuff

        EditText userinputfieldview = findViewById(R.id.newTaskField);

        View buttonView = findViewById(R.id.buttonAddTask);
        View.OnClickListener ocl = new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // 1. Grab the text the user has inputed into @id/newTaskField


                String userInputtedTask = userinputfieldview.getText().toString();

                // 2. add the string to the list of tasks arraylist

                listOfTasks.add(userInputtedTask);

                // 3. notify the adapter about the data change

                taskAdapter.notifyItemInserted(listOfTasks.size()-1);

                // 4. reset task field

                userinputfieldview.setText("");

                // 5. persistant save update

                saveItems();

            }
        };


        buttonView.setOnClickListener(ocl);




    }

    //file io system

    private File getDataFile(){
        // simple file list, nothing particular
        return new File(getFilesDir(), "data_java_ver.dat");
    }

    private void loadItems(){
        Log.i("filemanager:","file read attempt from "+getDataFile().getAbsolutePath().toString());
        try{
            this.listOfTasks = (ArrayList<String>) FileUtils.readLines(getDataFile(), Charset.defaultCharset());
            Log.i("filemanager:","file read from "+getDataFile().getAbsolutePath().toString());
        }
        catch (IOException ioException){
            ioException.printStackTrace();
        }

    }

    private void saveItems(){
        try{
            FileUtils.writeLines(getDataFile(), this.listOfTasks);
            Log.i("filemanager:","file saved in "+getDataFile().getAbsolutePath().toString());
        }
        catch (IOException ioException){
            ioException.printStackTrace();
        }
    }


}

File: app/src/main/java/com/neko/todolistapplication/MainActivity.kt
package com.neko.todolistapplication

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.EditText
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.apache.commons.io.FileUtils
import java.io.File
import java.io.IOException
import java.nio.charset.Charset

class MainActivity : AppCompatActivity() {



    var listOfTasks = mutableListOf<String>()
    lateinit var adapter: TaskListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val onLongClickListener = object :TaskListAdapter.OnLongClickListener {
            override fun onItemLongClicked(position: Int) {
                // remove the item from the list
                listOfTasks.removeAt(position)
                // update the adapter
                adapter.notifyDataSetChanged()

                saveItems()
            }
        }


        loadItems()



        val recyler_reference = findViewById<RecyclerView>(R.id.recyclerfield)
        adapter = TaskListAdapter(listOfTasks, onLongClickListener)
        recyler_reference.adapter = adapter
        recyler_reference.layoutManager = LinearLayoutManager(this)

        var inputTextField = findViewById<EditText>(R.id.newTaskField)

        // add button listener check
        findViewById<Button>(R.id.buttonAddTask).setOnClickListener {

            //grab the text from the @+id/newTaskField

            val userInputTask = inputTextField.text.toString()

            //add the string to our list of tasks: listOfTasks

            listOfTasks.add(userInputTask)

            //adapter calling for data update
            adapter.notifyItemInserted(listOfTasks.size - 1)

            //text field reset

            inputTextField.setText("")

            saveItems()
        }

    }



    //save data user inputted.

    //create a method to get the file

    fun getDataFile() : File {
        return File(filesDir, "data.lstxt")
    }

    fun loadItems(){
        try{
            listOfTasks = FileUtils.readLines(getDataFile(), Charset.defaultCharset())
        }
        catch (ioException: IOException){
            ioException.printStackTrace()
        }
    }

    fun saveItems(){
        try{
            FileUtils.writeLines(getDataFile(), listOfTasks)
        }
        catch (ioException: IOException){
            ioException.printStackTrace()
        }

    }
}